state-chain/chains/src/sol/api.rs

```diff
@@ -805,15 +806,126 @@ impl<Environment: SolanaEnvironment> SetGovKeyWithAggKey<SolanaCrypto> for Solan
 	}
 }
 
-impl<Env: 'static> RejectCall<Solana> for SolanaApi<Env> {
+impl<Environment: 'static + SolanaEnvironment> RejectCall<Solana> for SolanaApi<Environment> {
 	fn new_unsigned(
-		_deposit_details: <Solana as Chain>::DepositDetails,
-		_refund_address: <Solana as Chain>::ChainAccount,
-		_refund_amount: Option<<Solana as Chain>::ChainAmount>,
-		_asset: <Solana as Chain>::ChainAsset,
-		_deposit_fetch_id: Option<<Solana as Chain>::DepositFetchId>,
+		deposit_details: <Solana as Chain>::DepositDetails,
+		refund_address: <Solana as Chain>::ChainAccount,
+		refund_amount: Option<<Solana as Chain>::ChainAmount>,
+		asset: <Solana as Chain>::ChainAsset,
+		deposit_fetch_id: Option<<Solana as Chain>::DepositFetchId>,
 	) -> Result<Self, RejectError> {
-		todo!()
+		// Lookup environment variables
+		let agg_key = Environment::current_agg_key().map_err(|_| RejectError::Other)?;
+		let sol_api_environment = Environment::api_environment().map_err(|_| RejectError::Other)?;
+		let compute_price = Environment::compute_price().map_err(|_| RejectError::Other)?;
+		let durable_nonce = Environment::nonce_account().map_err(|_| RejectError::Other)?;
+
+		// We expect Deposit channels to provide a deposit_fetch_id and Vault swaps to not have one.
+		if matches!(deposit_details, VaultSwapOrDepositChannelId::VaultSwapAccount(_)) &&
+			deposit_fetch_id.is_some()
+		{
+			log_or_panic!("Invalid rejection: Vault swap ingress with a deposit_fetch_id");
+		} else if matches!(deposit_details, VaultSwapOrDepositChannelId::Channel(_)) &&
+			deposit_fetch_id.is_none()
+		{
+			log_or_panic!("Invalid rejection: Deposit channel ingress with no deposit_fetch_id");
+		}
+
+		let transaction = match (refund_amount, deposit_fetch_id, asset) {
+			// Vault swap with no refund amount doesn't require any action
+			(None, None, _) => return Err(RejectError::NotRequired),
+			// Just fetch from deposit channel - no refund amount specified
+			(None, Some(fetch_id), asset) => {
+				let fetch_params = FetchAssetParams { deposit_fetch_id: fetch_id, asset };
+				SolanaTransactionBuilder::fetch_from(
+					vec![fetch_params],
+					sol_api_environment,
+					agg_key,
+					durable_nonce,
+					compute_price,
+				)
+			},
```

Comment:
What does this do? Is this fetching from a channel that we want to reject, but since we don't know how much to reject, we just ingress it?

Can this case ever happen? The "no refund amount specified" I mean.

Reply:
Nice!